libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,696 @@
+use errors::UrsaCryptoError;
+
+use hash::{digest, DigestAlgorithm, Digest, sha2};
+use num_bigint::{BigInt, BigUint, RandBigInt, ToBigInt, Sign};
+use num_integer::Integer;
+use num_traits::identities::{One, Zero};
+use num_traits::{Num, Signed, ToPrimitive, Pow};
+use glass_pumpkin::{prime, safe_prime};
+use rand::rngs::OsRng;
+
+#[cfg(feature = "serialization")]
+use serde::ser::{Serialize, Serializer, Error as SError};
+
+#[cfg(feature = "serialization")]
+use serde::de::{Deserialize, Deserializer, Visitor, Error as DError};
+
+use std::error::Error;
+use std::fmt;
+use std::cmp::Ord;
+use std::cmp::Ordering;
+
+pub struct BigNumberContext;
+
+pub struct BigNumber {
+    bn: BigInt
+}
+
+macro_rules! prime_generation {
+    ($f:ident, $size:ident, $msg:expr) => {
+        match $f::new($size)?.to_bigint() {
+            Some(bn) => Ok(BigNumber { bn }),
+            None => Err(UrsaCryptoError::InvalidStructure($msg.to_string()))
+        }
+    };
+}
+
+macro_rules! prime_check {
+    ($f:ident, $value:expr, $msg:expr) => {
+        if $value.is_negative() {
+            Ok(false)
+        } else {
+            match $value.bn.to_biguint() {
+                Some(bn) => Ok($f::check(&bn)),
+                None => Err(UrsaCryptoError::InvalidStructure($msg.to_string()))
+            }
+        }
+    };
+}
+
+impl BigNumber {
+    pub fn new_context() -> Result<BigNumberContext, UrsaCryptoError> {
+        Ok(BigNumberContext{})
+    }
+
+    pub fn new() -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber {
+            bn: BigInt::zero()
+        })
+    }
+
+    pub fn generate_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        prime_generation!(prime, size, "Unable to generate prime")
+    }
+
+    pub fn generate_safe_prime(size: usize) -> Result<BigNumber, UrsaCryptoError> {
+        prime_generation!(safe_prime, size, "Unable to generate safe prime")
+    }
+
+    pub fn generate_prime_in_range(start: &BigNumber, end: &BigNumber) -> Result<BigNumber, UrsaCryptoError> {
+        let mut res;
+        let mut iteration = 0;
+        let mut rng = OsRng::new()?;
+        let mut start = match start.bn.to_biguint() {
+            Some(bn) => bn,
+            None => return Err(UrsaCryptoError::InvalidStructure(format!("Invalid number for 'start': {:?}", start)))
+        };
+        let mut end = match end.bn.to_biguint() {
+            Some(bn) => bn,
+            None => return Err(UrsaCryptoError::InvalidStructure(format!("Invalid number for 'end': {:?}", end)))
+        };
+
+        if start > end {
+            let temp = start;
+            start = end.clone();
+            end = temp;
+        }
+
+        loop {
+            res = rng.gen_biguint_range(&start, &end);
+            res |= BigUint::one();
```

Comment:
`BigUint::one()` will create a new `BigUint` each time, why not call it once outside the loop, assign the result to a variable and use the variable inside the loop?

Reply:
When I view the `asm` code using cargo-asm it appears to be optimized and only create it once anyway.